### PR TITLE
Extract theme folder scanning into an external class FolderThemeScanner

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Extract theme folder scanning into an external class `FolderThemeScanner` which will help the modules a lot, add a few interfaces on collections, remove `actionGetMailThemeFolder` hook
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | ~
| How to test?  | No QA required as the automatic tests validate the refacto didn't break anything

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13869)

### CODE REVIEW

File: src/Core/MailTemplate/FolderThemeCatalog.php [hunk 1]
```diff
@@ -61,6 +56,7 @@ public function __construct($mailThemesFolder, HookDispatcherInterface $hookDisp
     {
         $this->mailThemesFolder = $mailThemesFolder;
         $this->hookDispatcher = $hookDispatcher;
+        $this->scanner = new FolderThemeScanner();
```
Comment (PierreRambaud) [inline src/Core/MailTemplate/FolderThemeCatalog.php]: Must be injected in the constructor

Comment (jolelievre) [reply src/Core/MailTemplate/FolderThemeCatalog.php]: Yep, legit! Done

File: install-dev/data/xml/tab.xml [hunk 2]
```diff
@@ -204,9 +204,13 @@
                 <tab id="Themes_Catalog" id_parent="Look_feel" active="1" hide_host_mode="0" icon="">
                     <class_name>AdminThemesCatalog</class_name>
                 </tab>
-                <tab id="Mail_Theme" id_parent="Look_feel" active="1" hide_host_mode="0" icon="">
-                  <class_name>AdminMailTheme</class_name>
+                <tab id="Mail_Theme_Parent" id_parent="Look_feel" active="1" hide_host_mode="0" icon="">
+                  <class_name>AdminMailThemeParent</class_name>
                 </tab>
+                    <tab id="Mail_Theme" id_parent="Mail_Theme_Parent" active="1" hide_host_mode="0" icon="">
```
Comment (PierreRambaud) [inline install-dev/data/xml/tab.xml:210]: weird indent

Comment (jolelievre) [reply install-dev/data/xml/tab.xml:210]: Yes, it is actually on purpose to "illustrate" the parent/child association Other tabs are indented the same way

File: src/Core/MailTemplate/Theme.php [hunk 3]
```diff
@@ -47,6 +48,7 @@ class Theme implements ThemeInterface
     public function __construct($name)
     {
         $this->name = $name;
+        $this->layouts = new LayoutCollection();
```
Comment (PierreRambaud) [inline src/Core/MailTemplate/Theme.php:51]: any reason to add this here? You can use `setLayouts` instead, or use constructor like 

```php
public function __construct($name, $layouts = null)
{
    $this->layouts = $layouts === null ? new LayoutCollection() : $layouts;
}
```

Even if i'm not a big fan of this solution

File: src/Core/MailTemplate/ThemeCollection.php [hunk 4]
```diff
@@ -40,4 +40,21 @@ protected function getType()
...
+
+    /**
+     * @param string $themeName
+     *
+     * @return ThemeInterface|null
+     */
+    public function getByName($themeName)
+    {
+        /** @var ThemeInterface $theme */
+        foreach ($this as $theme) {
```
Comment (PierreRambaud) [inline src/Core/MailTemplate/ThemeCollection.php]: Try to no use magic as must as possible.
```suggestion
        foreach ($this->getValues() as $theme) {
```

File: src/Core/MailTemplate/ThemeCollection.php [hunk 5]
```diff
@@ -40,4 +40,21 @@ protected function getType()
...
+            if ($themeName === $theme->getName()) {
```
Comment (PierreRambaud) [inline src/Core/MailTemplate/ThemeCollection.php]: use https://www.doctrine-project.org/api/collections/latest/Doctrine/Common/Collections/ArrayCollection.html#method_filter instead ?

File: src/Core/MailTemplate/Theme.php (hunk 3 above)
Comment (jolelievre) [reply src/Core/MailTemplate/Theme.php:51]: The main reason was to init the field, so you can do `$theme->getLayouts()->count()` even with an empty theme (no need to check for null)

File: src/Core/MailTemplate/ThemeCollection.php (hunk 5 above)
Comment (jolelievre) [reply src/Core/MailTemplate/ThemeCollection.php]: Oh I didn't think we could benefit from this method, great idea!
But in this case it's not relevant, filter returns a collection and we only want one item returned in this method

File: src/Core/MailTemplate/ThemeCollection.php (hunk 4 above)
Comment (jolelievre) [reply src/Core/MailTemplate/ThemeCollection.php]: Done

File: src/Core/MailTemplate/ThemeCollection.php (hunk 5 above)
Comment (PierreRambaud) [reply src/Core/MailTemplate/ThemeCollection.php]: Combine methods:

```php
$this->filter($filterCondition)->first();
```

**Comment by matks** [comment]
Restarting Travis build as we got (again) some memory corruption

File: src/Core/MailTemplate/ThemeCollection.php (hunk 5 above)
Comment (jolelievre) [reply src/Core/MailTemplate/ThemeCollection.php]: Ok, done!

**Comment by jolelievre** [comment]
> Restarting Travis build as we got (again) some memory corruption

Thanks, but I just rebased the branch (along with a little modif) to restart the build
Fingers crossed!

**Review by PierreRambaud — APPROVED** [review]

**Pull request merged**